### PR TITLE
Adaptive time stepping through wizardry

### LINKED ISSUE #189: Adaptive time stepping
Might be interesting to give adaptive time-stepping a try. @christophernhill says it didn't work super well last time they tried it in the MITgcm but that might have been a long time ago. I suspect for most MITgcm runs, which act on large horizontal domains, the time step is severely limited by vertical motions which may not be true for all Oceananigans simulations. Not sure if this changes anything.

I suspect you just calculate the time step using
```julia
C = 1  # Maximum Courant number. Probably 1 for our explicit AB2 time-stepping scheme.

Δt_x = C*Δx / maximum(abs.(u))
Δt_y = C*Δy / maximum(abs.(v))
Δt_z = C*Δz / maximum(abs.(w))

Δt = min(Δt_x, Δt_y, Δt_z)
```

Even if we don't use adaptive time-stepping, it might be a useful `CFLChecker` diagnostic that prints warnings when CFL is being violated.

### PR DESCRIPTION
**Pull request by ali-ramadhan**
Copy pasting from @glwagner's fork.

Adaptive time stepping with a `TimeStepWizard` that computes time steps for you.

I just need to modify `time_step!` with a flag like `first_step_Euler=true` to take forward Euler steps at first iteration and when changing the time step, but with the ability to turn it off for tests where we actually don't want to do this. 

I'll add some more docstrings and a couple of tests.

cc @sandreza 

Resolves #189

### CODE REVIEW

**Comment by johncmarshall54** [comment]
Guys, is this a good idea? John

**Comment by glwagner** [comment]
What are the concerns?

**Comment by ali-ramadhan** [comment]
> Guys, is this a good idea? John

It's an optional set of utilities we can use to choose a time step. Default will still be constant dt.

I hear it's easy to get burned by adaptive time-stepping, especially if taking time steps close to what's allowed by CFL.

But I think it'll be especially useful for the seasonal cycle simulations where instead of running with dt = 3 seconds, we can set a maximum of dt = 3 seconds and let the adaptive time stepping ensure we don't blow up just because we were suddenly restricted to dt = 2.9 seconds by CFL. Or we can always take time steps that are e.g. 0.2*max_dt to ensure we're significantly below CFL.

**Comment by glwagner** [comment]
Adaptive time-stepping is useful for a wide class of problems. The general strategy is to use the CFL number to compute a stable time-step. This is especially useful in problems with lots of time variability --- for example, LES of a seasonal cycle. In addition, the CFL computation can be used to stabilize the spin-up of a model, during which a transition to turbulence occurs.

Some complexity is introduced by the fact that we use Adams-Bashforth time stepping (as opposed to a RK-method, which is more common and more stable), since it means we need to re-initialize the "previous source terms" with a forward Euler step when the time step is changed.

I have been using adaptive time-stepping in my work, which along with a strategy for handling model spinup / transition to turbulence, has proven to be an important practical feature that makes running experimental simulations under different conditions much easier, since all I need to do is set my boundary conditions, and let the model run. I am then able to start dozens of LES per day (if I want to) without having to choose my time step size through a mistake-prone and time-consuming trial-and-error process --- and I am guaranteed not to waste resources with a too-short time-step.

**Comment by glwagner** [comment]
> I hear it's easy to get burned by adaptive time-stepping, especially if taking time steps close to what's allowed by CFL.

Adaptive time-stepping is useful for avoiding "getting burned", since you can ensure that your time-step is always stable by choosing a conservative CFL number.

> Or we can always take time steps that are e.g. 0.2*max_dt to ensure we're significantly below CFL.

I don't understand this comment. The `TimeStepWizard` has a parameter `cfl`, which sets `dt` to achieve a certain CFL number. If you set `cfl=0.1`, the time-step is chosen so that `CFL=0.2`. The parameter `max_dt` is a user-defined parameter that limits `dt` to some maximum number.

**Comment by glwagner** [comment]
> I just need to modify time_step! with a flag like first_step_Euler=true to take forward Euler steps at first iteration and when changing the time step, but with the ability to turn it off for tests where we actually don't want to do this.

This is a good idea. You could also call it `initialize_with_Euler`.

Another possibility is to make the `time_step` function more modular, which would make it easy to define a new non-initializing variant of `time_step!`with a special name (rather than using a flag).

File: src/time_step_utils.jl [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+Compute the maximum allowable time step for a `model` by the Courant–Friedrichs–Lewy (CFL)
+condition given a Courant number `cfl`. Will never return a time step greater than `max_Δt`.
+"""
+function cfl_Δt(model, cfl, max_Δt)
+    τ = Δmin(model.grid) / Umax(model)
+    return min(max_Δt, cfl*τ)
+end
+
+"""
```
Comment (glwagner) [inline src/time_step_utils.jl]: This function is superfluous and probably should be omitted from the PR.

File: src/time_step_utils.jl [hunk 2]
```diff
@@ -0,0 +1,104 @@
+Δmin(grid::RegularCartesianGrid) = min(grid.Δx, grid.Δy, grid.Δz)
+
+function Umax(model)
```
Comment (glwagner) [inline src/time_step_utils.jl]: This function is not used and should be deleted.

**Comment by ali-ramadhan** [comment]
> I don't understand this comment. The `TimeStepWizard` has a parameter `cfl`, which sets `dt` to achieve a certain CFL number. If you set `cfl=0.1`, the time-step is chosen so that `CFL=0.2`. The parameter `max_dt` is a user-defined parameter that limits `dt` to some maximum number.

Sorry `max_dt` wasn't a reference to the code. The point was that you can remain as stable as you want to be.

File: src/time_step_utils.jl [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    w = model.velocities.w.data.parent
+
+    u_max = maximum(u)
+    v_max = maximum(v)
+    w_max = maximum(w)
+
+    max(u_max, v_max, w_max)
+end
+
+"""
```
Comment (glwagner) [inline src/time_step_utils.jl]: This function is not used and should be deleted.

**Comment by johncmarshall54** [comment]
Ali and Greg, thanks for your comments. I found the exchange very
enlightening. John

**Comment by christophernhill** [comment]
Hi All - this should definitely be useful for the sort of ensemble of
LES parameter/background state
sweep runs that Greg mentions

**Comment by glwagner** [comment]
@ali-ramadhan, I think this is ready to merge.

**Pull request merged**